keychain_unix.cpp

```diff
@@ -76,28 +76,60 @@ static DesktopEnvironment detectDesktopEnvironment() {
     return DesktopEnv_Other;
 }
 
+static bool isKwallet5Available()
+{
+    if (!QDBusConnection::sessionBus().isConnected())
+        return false;
+
+    org::kde::KWallet iface(
+        QLatin1String("org.kde.kwalletd5"),
+        QLatin1String("/modules/kwalletd5"),
+        QDBusConnection::sessionBus());
+
+    // At this point iface.isValid() can return false even though the
+    // interface is activatable by making a call. Hence we check whether
+    // a wallet can be opened.
+
+    iface.setTimeout(500);
+    QDBusMessage reply = iface.call(QStringLiteral("networkWallet"));
+    return reply.type() == QDBusMessage::ReplyMessage;
+}
+
 static KeyringBackend detectKeyringBackend()
 {
-    /* Libsecret unifies access to KDE and GNOME
-     * password services. */
-    if (LibSecretKeyring::isAvailable()) {
-        return Backend_LibSecretKeyring;
-    }
+    /* The secret service dbus api, accessible through libsecret, is supposed
+     * to unify password services.
+     *
+     * Unfortunately at the time of Kubuntu 18.04 the secret service backend
+     * in KDE is gnome-keyring-daemon - using it has several complications:
+     * - the default collection isn't opened on session start, so users need
+     *   to manually unlock it when the first application uses it
+     * - it's separate from the kwallet5 keyring, so switching to it means the
+     *   existing keyring data can't be accessed anymore
+     *
+     * Thus we still prefer kwallet backends on KDE even if libsecret is
+     * available.
+     */
 
     switch (detectDesktopEnvironment()) {
     case DesktopEnv_Kde4:
         return Backend_Kwallet4;
-        break;
+
     case DesktopEnv_Plasma5:
-        return Backend_Kwallet5;
-        break;
-        // fall through
+        if (isKwallet5Available()) {
+            return Backend_Kwallet5;
+        }
+        // fallthrough
     case DesktopEnv_Gnome:
     case DesktopEnv_Unity:
     case DesktopEnv_Xfce:
     case DesktopEnv_Other:
     default:
-         if ( GnomeKeyring::isAvailable() ) {
+        if (LibSecretKeyring::isAvailable()) {
+            return Backend_LibSecretKeyring;
+        }
+
+        if ( GnomeKeyring::isAvailable() ) {
             return Backend_GnomeKeyring;
         } else {
             return Backend_Kwallet4;
```

Comment:
Please change back to Backend_Kwallet5

So this would cover the use case that we are running Plasma 5,  but neither isKwallet5Available() nor LibSecretKeyring::isAvailable() (because  the application just started at startup before kwallet5d is running)

Reply:
Right, that might happen during startup. So:
 
Plasma: Try kwallet5, try libsecret, try gnome-keyring, fallback kwallet5
Gnome and others: Try libsecret, try gnome-keyring, try kwallet5, fallback gnome-keyring

?